File: CoreVia/CoreVia/Features/Social/Models/SocialModels.swift
import Foundation

// MARK: - Social Post

struct SocialPost: Identifiable, Codable {
    let id: String
    let userId: String
    let postType: String  // workout, meal, progress, achievement, general
    let content: String?
    let imageUrl: String?
    let workoutId: String?
    let foodEntryId: String?
    var likesCount: Int
    var commentsCount: Int
    let isPublic: Bool
    let createdAt: Date
    let updatedAt: Date

    // Extra
    var author: PostAuthor?
    var isLiked: Bool

    enum CodingKeys: String, CodingKey {
        case id, content
        case userId = "user_id"
        case postType = "post_type"
        case imageUrl = "image_url"
        case workoutId = "workout_id"
        case foodEntryId = "food_entry_id"
        case likesCount = "likes_count"
        case commentsCount = "comments_count"
        case isPublic = "is_public"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case author
        case isLiked = "is_liked"
    }
}

struct PostAuthor: Codable {
    let id: String
    let name: String
    let profileImageUrl: String?
    let userType: String

    enum CodingKeys: String, CodingKey {
        case id, name
        case profileImageUrl = "profile_image_url"
        case userType = "user_type"
    }
}

// MARK: - Feed Response

struct FeedResponse: Codable {
    let posts: [SocialPost]
    let total: Int
    let page: Int
    let pageSize: Int
    let hasMore: Bool

    enum CodingKeys: String, CodingKey {
        case posts, total, page
        case pageSize = "page_size"
        case hasMore = "has_more"
    }
}

// MARK: - Post Create

struct CreatePostRequest: Codable {
    let postType: String
    let content: String?
    let workoutId: String?
    let foodEntryId: String?
    let isPublic: Bool

    enum CodingKeys: String, CodingKey {
        case content
        case postType = "post_type"
        case workoutId = "workout_id"
        case foodEntryId = "food_entry_id"
        case isPublic = "is_public"
    }
}

// MARK: - Comment

struct PostComment: Identifiable, Codable {
    let id: String
    let postId: String
    let userId: String
    let content: String
    let createdAt: Date
    var author: CommentAuthor?

    enum CodingKeys: String, CodingKey {
        case id, content, author
        case postId = "post_id"
        case userId = "user_id"
        case createdAt = "created_at"
    }
}

struct CommentAuthor: Codable {
    let id: String
    let name: String
    let profileImageUrl: String?

    enum CodingKeys: String, CodingKey {
        case id, name
        case profileImageUrl = "profile_image_url"
    }
}

// MARK: - Achievement

struct Achievement: Identifiable, Codable {
    let id: String
    let userId: String
    let achievementType: String
    let title: String
    let description: String?
    let iconUrl: String?
    let earnedAt: Date

    enum CodingKeys: String, CodingKey {
        case id, title, description
        case userId = "user_id"
        case achievementType = "achievement_type"
        case iconUrl = "icon_url"
        case earnedAt = "earned_at"
    }
}

File: CoreVia-Android/app/src/main/java/life/corevia/app/data/models/SocialModels.kt
package life.corevia.app.data.models

import com.google.gson.annotations.SerializedName

// ─── Social Models ──────────────────────────────────────────────────────────
// iOS: SocialManager.swift → Post, Comment, Achievement

// ─── Author nested objects (backend qaytarır) ───────────────────────────────

data class PostAuthor(
    val id: String? = null,
    val name: String? = null,
    @SerializedName("profile_image_url") val profileImageUrl: String? = null,
    @SerializedName("user_type") val userType: String? = null
)

data class CommentAuthor(
    val id: String? = null,
    val name: String? = null,
    @SerializedName("profile_image_url") val profileImageUrl: String? = null
)

// ─── Post ───────────────────────────────────────────────────────────────────

data class SocialPost(
    val id: String,
    @SerializedName("user_id")      val userId: String,
    val content: String? = null,
    @SerializedName("image_url")    val imageUrl: String? = null,
    @SerializedName("post_type")    val postType: String = "general",
    @SerializedName("likes_count")  val likeCount: Int = 0,
    @SerializedName("comments_count") val commentCount: Int = 0,
    @SerializedName("is_liked")     val isLiked: Boolean = false,
    @SerializedName("is_public")    val isPublic: Boolean = true,
    @SerializedName("workout_id")   val workoutId: String? = null,
    @SerializedName("food_entry_id") val foodEntryId: String? = null,
    @SerializedName("created_at")   val createdAt: String,
    @SerializedName("updated_at")   val updatedAt: String? = null,
    // Backend nested author object
    val author: PostAuthor? = null
) {
    // Compatibility helpers — backend author obj-dan oxu
    val userName: String? get() = author?.name
    val userImageUrl: String? get() = author?.profileImageUrl
}

// ─── Comment ────────────────────────────────────────────────────────────────

data class SocialComment(
    val id: String,
    @SerializedName("user_id")      val userId: String,
    @SerializedName("post_id")      val postId: String? = null,
    val content: String,
    @SerializedName("created_at")   val createdAt: String,
    // Backend nested author object
    val author: CommentAuthor? = null
) {
    val userName: String? get() = author?.name
    val userImageUrl: String? get() = author?.profileImageUrl
}

// ─── Create Requests ────────────────────────────────────────────────────────

data class CreatePostRequest(
    val content: String,
    @SerializedName("post_type") val postType: String = "general",
    @SerializedName("workout_id") val workoutId: String? = null,
    @SerializedName("food_entry_id") val foodEntryId: String? = null,
    @SerializedName("is_public") val isPublic: Boolean = true
)

data class CreateCommentRequest(
    val content: String
)

// ─── Feed Response Wrapper ──────────────────────────────────────────────────

data class FeedResponse(
    val posts: List<SocialPost> = emptyList(),
    val total: Int = 0,
    val page: Int = 1,
    @SerializedName("page_size") val pageSize: Int = 20,
    @SerializedName("has_more") val hasMore: Boolean = false
)

// ─── Achievement ────────────────────────────────────────────────────────────

data class Achievement(
    val id: String,
    @SerializedName("achievement_type") val achievementType: String? = null,
    val title: String,
    val description: String,
    val icon: String? = null,
    @SerializedName("is_unlocked") val isUnlocked: Boolean = false,
    @SerializedName("earned_at") val earnedAt: String? = null,
    @SerializedName("unlocked_at") val unlockedAt: String? = null,
    val progress: Double = 0.0,
    val target: Int = 1,
    val current: Int = 0
)

// ─── User Profile Summary ───────────────────────────────────────────────────

data class UserProfileSummary(
    val id: String,
    val name: String,
    @SerializedName("profile_image_url") val profileImageUrl: String? = null,
    @SerializedName("user_type")   val userType: String,
    val bio: String? = null,
    @SerializedName("is_following") val isFollowing: Boolean = false,
    @SerializedName("followers_count") val followerCount: Int = 0,
    @SerializedName("following_count") val followingCount: Int = 0,
    @SerializedName("posts_count")  val postCount: Int = 0
)
